Always show the parent trails in the tree view

Toflar opened the pull request:
| Q                | A
| -----------------| ---
| Fixed issues     | Fixes https://github.com/contao/contao/issues/3242

This PR is a first shot at trying to display the root elements of page mounts (or any other tree view).

Before (until Contao 4.12.x):

<img width="1213" alt="Bildschirmfoto 2021-08-27 um 14 10 00" src="https://user-images.githubusercontent.com/481937/131125543-448f18d7-c59f-434a-b826-b47fd2a5487a.png">

With this PR:

<img width="1212" alt="Bildschirmfoto 2021-08-27 um 14 09 05" src="https://user-images.githubusercontent.com/481937/131125575-f820eac1-8a86-43d1-8c6a-deaa28ae899c.png">

It not only shows the trail but also ensures a consistent sorting 😎

Toflar committed:
Implemented displaying the parent trails in tree views

Toflar commented on core-bundle/src/Resources/contao/drivers/DC_Table.php:
```diff
@@ -6204,6 +6180,75 @@ protected function formatGroupHeader($field, $value, $mode, $row)
+		}
+
+		// Sort the root based on the sorting of the root trails if sorting is enabled
+		if ($this->Database->fieldExists('sorting', $table))
```
This here ensures, the tree views are consistent according to the sorting of the topmost tree element.

Toflar commented on core-bundle/src/Resources/contao/library/Contao/Database.php:
```diff
@@ -572,6 +572,12 @@ public function getParentRecords($intId, $strTable)
 
 		while ($objPages->next())
 		{
+			// TODO: Find a better solution for this? Why does this method include the ID itself?
```
I'm not sure about this but why does `getParentRecords()` return an array including itself? If it says parent records? Can we change that?

fritzmg replied:
This appears to be out of convenience, as it is mostly used to check whether a specific node (or list of nodes) is part of a specific sub tree, including itself. While the method's name is misleading, we cannot change it for BC reasons. We could introduce a third parameter `$blnIncludeSelf = true` …

Toflar replied:
Implemented in b559a04e4362af8ee7271d6cb10b97c3756e6c84.

Toflar committed:
Improve Database::getParentRecords()

Toflar commented:
This is ready to be reviewed. While I was on it, I also fixed `Database::getParentRecords()` to be limited to 10 levels. While it might be enough in most cases, I don't see why we should not add recursion here to fix this properly :)

leofeyer commented:
> While it might be enough in most cases, I don't see why we should not add recursion here to fix this properly :)

It seems that it has been enough in **all** cases in the last 15 years – at least it was never a (reported) issue. So I doubt that it is necessary to change it now. 🤷‍♂️ But since the way you have implemented it does not hurt performance, I don't mind.

leofeyer commented:
The paste buttons should not be visible, should they?

<img width="1212" alt="" src="https://user-images.githubusercontent.com/1192057/132825269-b236d44f-be84-42ad-8195-a6163392c050.png">

Toflar committed:
CS

Toflar committed:
Do not show any operations for root trail tree elements

Toflar committed:
Do not show the breadcrumb link for pages in the trail

Toflar committed:
Show the trail also in the bread crumb

Toflar committed:
Make display of root trails entirely optional and disable when breadcrumb is used

Toflar commented:
Fixed. There are now no operations anymore for parent trails.
The PR is also updated:

* The feature is now disabled by default for BC, other existing trees. The setting is currently named `showRootTrails`.
* I've fixed the breadcrumb to also show the parent trail
* The node filtering (limiting the tree view) is now only enabled on links which are not part of the trail
* `showRootTrails` is dynamically disabled if you use the breadcrumb now because having a breadcrumb and still showing the trail in the tree makes no sense.

ausi commented on core-bundle/src/Resources/contao/library/Contao/Database.php:
```diff
@@ -559,23 +559,35 @@ public function getChildRecords($arrParentIds, $strTable, $blnSorting=false, $ar
+		if (\count($ids) === 10)
 		{
-			$arrReturn[] = $objPages->id;
+			$ids = array_merge($ids, $this->getParentRecords(end($ids), $strTable, $skipId));
```
```suggestion
			$ids = array_merge($ids, $this->getParentRecords(end($ids), $strTable, true));
```

As `end($ids)` is already in the `$ids` we should skip it in the recursion call I think, otherwise we could end up with duplicated ids.

Toflar replied:
Ah right :)

ausi commented on core-bundle/src/Resources/contao/library/Contao/Database.php:
```diff
@@ -559,23 +559,35 @@ public function getChildRecords($arrParentIds, $strTable, $blnSorting=false, $ar
+			$ids = array_values(array_flip($ids));
+		}
+
+		return $ids;
```
```suggestion
			unset($ids[array_search($intId, $ids)]);
		}

		return array_values($ids);
```
Nitpick ☺️

Toflar replied:
Yeah I just did the flipping because of performance reasons but the array is so small, it's probably irrelevant ;)

ausi reviewed (commented):
Apart from the duplicate IDs ✅

Toflar committed:
CS

Co-authored-by: Martin Auswöger <martin@auswoeger.com>

Toflar committed:
Improve recursion

Co-authored-by: Martin Auswöger <martin@auswoeger.com>

leofeyer committed:
Merge branch '4.x' into feature/visible-tree-trail

leofeyer reviewed (changes requested):
Something does not work as expected. These are the pagemounts:

<img width="879" alt="" src="https://user-images.githubusercontent.com/1192057/135269864-05a6f0ee-6bd5-4ad5-9cf3-f041e412bd61.png">

However, the user only sees one of the three selected pages:

<img width="874" alt="" src="https://user-images.githubusercontent.com/1192057/135269924-5cd7c24f-d192-4bc7-a017-d124a4749fc5.png">

When I remove one of the page mounts, the user sees another page but still not all pages:

<img width="555" alt="" src="https://user-images.githubusercontent.com/1192057/135269965-d9525d85-857d-4416-8aa1-b15e058d182e.png">

<img width="874" alt="" src="https://user-images.githubusercontent.com/1192057/135270017-b4136dc6-d589-4f57-90f6-676397625a4f.png">

Any idea what could be causing this?

leofeyer commented on core-bundle/src/Resources/contao/classes/Backend.php:
```diff
@@ -873,8 +873,8 @@ public static function addPagesBreadcrumb($strKey='tl_page_node')
-				// No link for the active page
-				if ($objPage->id == $intNode)
+				// No link for the active page or pages in the trail
+				if ($objPage->id == $intNode || (!$objUser->isAdmin && !$objUser->hasAccess($objPage->id, 'pagemounts')))
```
```suggestion
				if ($objPage->id == $intNode || !$objUser->hasAccess($objPage->id, 'pagemounts'))
```

`User::hasAccess()` already checks `isAdmin`, therefore this condition can possibly be simplified.

Toflar replied:
Right, I just copied the original code ;)

Toflar committed:
Simplify access check

Co-authored-by: Leo Feyer <github@contao.org>

Toflar committed:
Fixed sorting of root pages removing root nodes

Toflar committed:
CS

Toflar committed:
Do not render already rendered parents multiple times

Toflar commented:
Fixed and found another issue with root nodes having the same parent which is also fixed now

Toflar commented:
There's still a bug with rendering though. On it.

Toflar committed:
Fixed wrong indentation

Toflar commented:
Now you can review again :)

leofeyer commented:
Almost there. Here's the admin view:

<img width="315" alt="" src="https://user-images.githubusercontent.com/1192057/135289064-a6e109fc-33b9-4cad-93b0-cc879840c741.png">

And here is the user view:

<img width="344" alt="" src="https://user-images.githubusercontent.com/1192057/135289111-a847fb2d-ec40-4a11-803a-4d21b41af0da.png">

The order is not yet correct. Or does it depend on the order of the pagemounts?

Toflar commented:
The order matters for routing so it should be the same (hence this PR in the first place, so you can actually see, which root page is first and it must be consistent with the admin view). I'll try to reproduce your issue.

leofeyer commented:
Thank you.

Toflar committed:
Completely reworked concept

Toflar committed:
CS

Toflar commented:
Completely reworked the concept. It's now going top down and skipping stuff that should not be visible instead of going up from the visible elements which makes sorting a lot easier :)

asaage commented:
If the resulting site-structure is a tree anyway then the ordering of pagemounts in user/groups should in my opinion be totaly omited.

Toflar commented:
It's not ordered so nothing to change.

leofeyer commented on core-bundle/src/Resources/contao/drivers/DC_Table.php:
```diff
@@ -209,42 +208,7 @@ public function __construct($strTable, $arrModule=array())
-		{
-			$this->root = array_unique($GLOBALS['TL_DCA'][$this->strTable]['list']['sorting']['root']);
-		}
+		$this->initRoots();
```
Why did you make this a separate method? It is not used anywhere else.

Toflar replied:
Because it's nicer. We keep talking about refactoring the DC_Table so here's one step: Move the logic of deterimining the roots into its own method.

leofeyer committed:
CS

leofeyer reviewed (changes requested):
I think adding type safety might be a BC break, because we do not know for sure how third-party extensions use the hooks or extend the `DC_Table` class. I would therefore prefer not to require type safety.

Great job otherwise! 💪

Toflar commented:
I think it's a must because otherwise, proper `in_array()` checks are almost impossible. I don't think it's a BC break, existing comparisons should continue to work just fine.

leofeyer commented:
I am not worried about existing comparisons but rather that a child class sets `$this->root` with string IDs. Then the strict `in_array()` checks would fail (DC_Table line 3844).

leofeyer committed:
Make the checks non-strict again

leofeyer commented:
I have found another issue with the implementation: We now see the trail pages but we cannot click on them to limit the site structure to a parent node.

<img width="449" alt="" src="https://user-images.githubusercontent.com/1192057/135636249-d4712805-e2c6-4b81-be2e-1debb55b4d42.png">

Wouldn't it be handy if we could click on "Modules" to only show this subtree?

**Edit:** This is a different issue that we should handle in a separate PR. Never mind.

Toflar committed:
Merge pull request #6 from leofeyer/fix/remove-strict-types

Make the checks non-strict again

leofeyer committed:
Fix the vertical alignment of the pages without link

leofeyer committed:
Remove a left-over int cast

leofeyer committed:
Handle the case that array_search() returns false

leofeyer committed:
Merge branch '4.x' into feature/visible-tree-trail

leofeyer committed:
Run the Gulp task

leofeyer commented:
Thank you @Toflar.
